File: ASCOM.Utilities/ASCOM.Utilities/AlpacaSupport/Entity Classes/ConfiguredDevice.vb
Imports System.Runtime.InteropServices
Imports ASCOM.Utilities.Interfaces

''' <summary>
''' Description of a single ASCOM device as returned within the <see cref="AlpacaDevice"/> class.
''' </summary>
<Guid("AAB2619C-69FC-4717-8F01-10228D59E99E"),
ComVisible(True),
ClassInterface(ClassInterfaceType.None)>
Public Class ConfiguredDevice
    Implements IConfiguredDevice

    ' State variables
    Dim deviceNameValue, deviceTypeValue, uniqueIDValue As String

    ''' <summary>
    ''' Initialises the class with default values
    ''' </summary>
    Public Sub New()
        ' Initialise all strings to empty values
        FixNullStrings()
    End Sub

    ''' <summary>
    ''' Initialise the device name, device type, device number and ASCOM device unique ID
    ''' </summary>
    ''' <paramname="deviceName">ASCOM device name</param>
    ''' <paramname="deviceType">ASCOM device type</param>
    ''' <paramname="deviceNumber">Alpaca API device number</param>
    ''' <paramname="uniqueID">ASCOM device unique ID</param>
    Friend Sub New(ByVal deviceName As String, ByVal deviceType As String, ByVal deviceNumber As Integer, ByVal uniqueID As String)
        deviceNameValue = deviceName
        deviceTypeValue = deviceType
        Me.DeviceNumber = deviceNumber
        uniqueIDValue = uniqueID

        ' Ensure that none of the strings are null
        FixNullStrings()
    End Sub

    ''' <summary>
    ''' ASCOM device name
    ''' </summary>
    Public Property DeviceName As String Implements IConfiguredDevice.DeviceName
        Get
            Return deviceNameValue
        End Get
        Set(value As String)
            deviceNameValue = value
            ' Protect against null values being assigned
            If String.IsNullOrEmpty(deviceNameValue) Then deviceNameValue = ""
        End Set
    End Property

    ''' <summary>
    ''' ASCOM device type
    ''' </summary>
    Public Property DeviceType As String Implements IConfiguredDevice.DeviceType
        Get
            Return deviceTypeValue
        End Get
        Set(value As String)
            deviceTypeValue = value
            ' Protect against null values being assigned
            If String.IsNullOrEmpty(deviceTypeValue) Then deviceTypeValue = ""
        End Set
    End Property

    ''' <summary>
    ''' Device number used to access the device through the Alpaca API
    ''' </summary>
    Public Property DeviceNumber As Integer Implements IConfiguredDevice.DeviceNumber

    ''' <summary>
    ''' ASCOM device unique ID
    ''' </summary>
    Public Property UniqueID As String Implements IConfiguredDevice.UniqueID
        Get
            Return uniqueIDValue
        End Get
        Set(value As String)
            uniqueIDValue = value
            ' Protect against null values being assigned
            If String.IsNullOrEmpty(uniqueIDValue) Then uniqueIDValue = ""
        End Set
    End Property

    ''' <summary>
    ''' Make sure that all properties return a string i.e. that they do not return "null" or an unhelpful empty string
    ''' </summary>
    Private Sub FixNullStrings()
        If String.IsNullOrEmpty(deviceNameValue) Then deviceNameValue = ""
        If String.IsNullOrEmpty(deviceTypeValue) Then deviceTypeValue = ""
        If String.IsNullOrEmpty(uniqueIDValue) Then uniqueIDValue = ""
    End Sub
End Class

File: ASCOM.Utilities/ASCOM.Utilities/AlpacaSupport/Entity Classes/ConfiguredDevice.cs
using System;
using System.Runtime.InteropServices;
using ASCOM.Utilities.Interfaces;

namespace ASCOM.Utilities
{

    /// <summary>
/// Data object describing a single ASCOM device as returned within the <see cref="AlpacaDevice"/> class.
/// </summary>
    [Guid("AAB2619C-69FC-4717-8F01-10228D59E99E")]
    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.None)]
    public class ConfiguredDevice : IConfiguredDevice
    {

        // State variables
        private string deviceNameValue, deviceTypeValue, uniqueIDValue;

        /// <summary>
    /// Initialises the class with default values
    /// </summary>
        public ConfiguredDevice()
        {
            // Initialise all strings to empty values
            FixNullStrings();
        }

        /// <summary>
    /// Initialise the device name, device type, device number and ASCOM device unique ID
    /// </summary>
    /// <param name="deviceName">ASCOM device name</param>
    /// <param name="deviceType">ASCOM device type</param>
    /// <param name="deviceNumber">Alpaca API device number</param>
    /// <param name="uniqueID">ASCOM device unique ID</param>
        internal ConfiguredDevice(string deviceName, string deviceType, int deviceNumber, string uniqueID)
        {
            deviceNameValue = deviceName;
            deviceTypeValue = deviceType;
            DeviceNumber = deviceNumber;
            uniqueIDValue = uniqueID;

            // Ensure that none of the strings are null
            FixNullStrings();
        }

        /// <summary>
    /// ASCOM device name
    /// </summary>
        public string DeviceName
        {
            get
            {
                return deviceNameValue;
            }
            set
            {
                deviceNameValue = value;
                // Protect against null values being assigned
                if (string.IsNullOrEmpty(deviceNameValue))
                    deviceNameValue = "";
            }
        }

        /// <summary>
    /// ASCOM device type
    /// </summary>
        public string DeviceType
        {
            get
            {
                return deviceTypeValue;
            }
            set
            {
                deviceTypeValue = value;
                // Protect against null values being assigned
                if (string.IsNullOrEmpty(deviceTypeValue))
                    deviceTypeValue = "";
            }
        }

        /// <summary>
    /// Device number used to access the device through the Alpaca API
    /// </summary>
        public int DeviceNumber { get; set; }

        /// <summary>
    /// ASCOM device unique ID
    /// </summary>
        public string UniqueID
        {
            get
            {
                return uniqueIDValue;
            }
            set
            {
                uniqueIDValue = value;
                // Protect against null values being assigned
                if (string.IsNullOrEmpty(uniqueIDValue))
                    uniqueIDValue = "";
            }
        }

        /// <summary>
    /// Make sure that all properties return a string i.e. that they do not return "null" or an unhelpful empty string
    /// </summary>
        private void FixNullStrings()
        {
            if (string.IsNullOrEmpty(deviceNameValue))
                deviceNameValue = "";
            if (string.IsNullOrEmpty(deviceTypeValue))
                deviceTypeValue = "";
            if (string.IsNullOrEmpty(uniqueIDValue))
                uniqueIDValue = "";
        }
    }
}